File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.py
def solution(n, tops):
    MOD = 10007
    answer = 0
    
    not_right = [0] * n
    right = [0] * n
    
    not_right[0] = 3 if tops[0] == 1 else 2
    right[0] = 1
    
    for i in range(1, n):
        if tops[i] == 1: # 위에 삼각형 붙어있는 경우
            not_right[i] = (3 * not_right[i-1] + 2 * right[i-1]) % MOD
            right[i] = (not_right[i-1] + right[i-1]) % MOD
        else:
            not_right[i] = (2 * not_right[i-1] + right[i-1]) % MOD
            right[i] = (not_right[i-1] + right[i-1]) % MOD
    
    answer = (not_right[-1] + right[-1]) % MOD
    return answer

File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.java
class Solution {
    static final int MOD = 10007;
    
    public int solution(int n, int[] tops) {
        int answer = 0;
        
        int[] notRight = new int[n];
        int[] right = new int[n];
        
        notRight[0] = tops[0] == 1 ? 3 : 2;
        right[0] = 1;
        
        for (int i = 1; i < n; i++) {
            int nr = notRight[i-1];
            int r = right[i-1];
            
            if (tops[i] == 1) {
                notRight[i] = (3 * nr + 2 * r) % MOD;
            } else {
                notRight[i] = (2 * nr + 1 * r) % MOD;
            }
            right[i] = (1 * nr + 1 * r) % MOD;
        }
        
        answer = (notRight[n-1] + right[n-1]) % MOD;
        return answer;
    }
}